### PR TITLE
[tests] migrate lora tests for pipelines to use new mixins

### PR DESCRIPTION
**Pull request by sayakpaul**
* Similar to `tests/models/testing_utils/lora.py`, this PR introduces a `tests/pipelines/testing_utils/lora.py`. The newly introduced mixins follow the same principles as the existing ones. This PR only migrates the pipeline-level LoRA tests for Flux. Others will follow in a different PR.
* There are tests in `tests/lora/utils.py` there are better tested at the model-level. This PR migrates those from `tests/lora/utils.py` to `tests/models/testing_utils/lora.py`.
  * As a consequence of this the model-level testers have been updated to use the `LoRATesterMixin` from `tests/models/testing_utils/lora.py` which were previously being tested through pipelines in `tests/lora`.

### CODE REVIEW

File: tests/pipelines/flux/test_pipeline_flux.py [hunk 1]
```diff
@@ -356,6 +363,137 @@ class TestFluxPipelineMagCache(FluxPipelineTesterConfig, MagCacheTesterMixin):
     """MagCache tests for the Flux pipeline."""
 
 
+class TestFluxPipelineLoRA(FluxPipelineTesterConfig, LoraTesterMixin):
+    """LoRA tests for the Flux pipeline."""
+
+    def test_with_alpha_in_state_dict(self, tmp_path, denoiser_lora_config):
```
Comment (sayakpaul) [inline tests/pipelines/flux/test_pipeline_flux.py]: These are all very specific to Flux.

File: tests/lora/test_lora_layers_ace_step.py [hunk 2]
```diff
@@ -23,11 +23,7 @@
 from diffusers.pipelines.ace_step import AceStepConditionEncoder, AceStepPipeline
 from diffusers.utils.import_utils import is_peft_available
 
-from ..testing_utils import (
```
Comment (sayakpaul) [inline tests/lora/test_lora_layers_ace_step.py:26]: The changes in the `test_lora_layers_*.py` (except for Flux), are related to the fact that some tests didn't really belong at the pipeline-level and are better off in the model-level (`tests/models/testing_utils/lora.py`).

File: tests/models/transformers/test_models_transformer_sana.py [hunk 3]
```diff
@@ -113,3 +115,9 @@ def test_gradient_checkpointing_is_applied(self):
 
 class TestSanaTransformerAttention(SanaTransformerTesterConfig, AttentionTesterMixin):
     pass
+
+
+class TestSanaTransformerLoRA(SanaTransformerTesterConfig, LoraTesterMixin):
+    @pytest.mark.skipif(IS_GITHUB_ACTIONS, reason="Skipping test inside GitHub Actions environment")
```
Comment (DN6) [inline tests/models/transformers/test_models_transformer_sana.py:122]: Any idea what the underlying issue is here?

File: lora-tests-migration-plan.md [hunk 4]
```diff
@@ -0,0 +1,375 @@
+# Plan: migrate `tests/lora/` to pipeline-level LoRA tester mixins
```
Comment (DN6) [inline lora-tests-migration-plan.md]: Should remove

File: tests/models/testing_utils/lora.py [hunk 5]
```diff
@@ -235,6 +240,351 @@ def test_lora_adapter_wrong_metadata_raises_error(self, tmp_path):
...
+        output = model(**inputs_dict, return_dict=False)[0]
+        # Some models (e.g. Z-Image) return a list of per-sample tensors — flatten so the tensor comparisons in
+        # the tests below work regardless of the output layout.
+        if isinstance(output, (list, tuple)):
+            output = torch.cat([t.flatten() for t in output])
+        return output
+
+    @require_peft_version_greater("0.13.1")
+    @torch.no_grad()
+    def test_lora_low_cpu_mem_usage_with_injection(self):
```
Comment (DN6) [inline tests/models/testing_utils/lora.py]: Here it feels like we're testing PEFT functionality rather than diffusers?

File: tests/models/testing_utils/lora.py [hunk 6]
```diff
@@ -235,6 +240,351 @@ def test_lora_adapter_wrong_metadata_raises_error(self, tmp_path):
...
+        lora_bias_false_output = self._model_output(model, inputs_dict)
+        model.delete_adapters("adapter-1")
+
+        model.add_adapter(LoraConfig(**lora_config_kwargs, lora_bias=True), adapter_name="adapter-1")
+        torch.manual_seed(0)
+        lora_bias_true_output = self._model_output(model, inputs_dict)
+
+        assert not torch.allclose(original_output, lora_bias_false_output, atol=atol, rtol=rtol)
+        assert not torch.allclose(original_output, lora_bias_true_output, atol=atol, rtol=rtol)
+        assert not torch.allclose(lora_bias_false_output, lora_bias_true_output, atol=atol, rtol=rtol)
```
Comment (DN6) [inline tests/models/testing_utils/lora.py:370]: In this case, since `"init_lora_weights": False`, won't the weights that are created in the `bias=True` case be completely different from the the ones created with `bias=False` since no seed is set? We would expect them to be different regardless of the bias being present or not?

File: tests/pipelines/testing_utils/lora.py [hunk 7]
```diff
@@ -0,0 +1,1544 @@
...
+            lora_alpha=self.lora_alpha,
+            target_modules=self.denoiser_target_modules,
+            init_lora_weights=False,
+            use_dora=False,
+        )
+
+    def get_denoiser(self, pipe):
+        return pipe.transformer if hasattr(pipe, "transformer") else pipe.unet
+
+    def run_pipe(self, pipe, **extra_inputs):
```
Comment (DN6) [inline tests/pipelines/testing_utils/lora.py]: This can probably also live in BasePipelineOutputMixin

File: tests/pipelines/testing_utils/lora.py [hunk 8]
```diff
@@ -0,0 +1,1544 @@
...
+        torch.manual_seed(0)
+        return pipe(**inputs)[0]
+
+    def _lora_text_encoder_2(self, pipe):
+        """The pipeline's `text_encoder_2` when it exists and supports LoRA loading, else None."""
+        if "text_encoder_2" in self.pipeline_class._lora_loadable_modules:
+            return getattr(pipe, "text_encoder_2", None)
+        return None
+
+    def add_adapters_to_pipeline(self, pipe, text_lora_config=None, denoiser_lora_config=None, adapter_name="default"):
```
Comment (DN6) [inline tests/pipelines/testing_utils/lora.py]: I think we can make this more general and leverage `_lora_loadable_modules`. 

```python
# TE keys live here since the architecture for a TE is unlikely to vary across pipelines. 
TEXT_ENCODER_TARGET_MODULES = {
    "clip_text_model": ["q_proj", "k_proj", "v_proj", "out_proj"],
    "clip_text_model_with_projection": ["q_proj", "k_proj", "v_proj", "out_proj"],
}

@is_lora
@require_peft_backend
class BaseLoraTesterMixin(BasePipelineOutputMixin):
    lora_rank = 4
    lora_alpha = 4

    # denoiser can be overriden at pipeline test instance level since these keys could potentially change    across pipelines. 
    denoiser_target_modules = {
        "unet": ["to_q", "to_k", "to_v", "to_out.0"],
        "transformer": ["to_q", "to_k", "to_v", "to_out.0"],
    }

    def get_target_modules(self, name, module):
        """Return the LoRA `target_modules` for the pipeline component `name`."""
        if name in self.denoiser_target_modules:
            return self.denoiser_target_modules[name]
        if name.startswith("text_encoder"):
            model_type = module.config.model_type
            if model_type not in TEXT_ENCODER_TARGET_MODULES:
                raise ValueError(
                    f"No LoRA target modules registered for text encoder model_type={model_type!r}. "
                    f"Add an entry to TEXT_ENCODER_TARGET_MODULES in tests/pipelines/testing_utils/lora.py."
                )
            return TEXT_ENCODER_TARGET_MODULES[model_type]
        raise ValueError(f"Cannot determine LoRA target modules for pipeline component {name!r}.")

    def add_adapters_to_pipeline(self, pipe, components=None, adapter_name="default"):
        """Attach a LoRA adapter to the given components (default: all LoRA-loadable ones).

        Returns {component_name: module} for everything adapted, e.g. for passing to
        `save_lora_weights` via `_get_lora_state_dicts`.
        """
        components = components if components is not None else self.pipeline_class._lora_loadable_modules
        adapted = {}
        for name in components:
            module = getattr(pipe, name, None)
            if module is None:
                continue
            config = LoraConfig(
                r=self.lora_rank,
                lora_alpha=self.lora_alpha,
                target_modules=self.get_target_modules(name, module),
                init_lora_weights=False,
            )
            module.add_adapter(config, adapter_name=adapter_name)
            assert check_if_lora_correctly_set(module), f"LoRA not correctly set in {name}"
            adapted[name] = module
        return adapted
```

Usage: 

```python
# Before
def test_simple_inference_with_text_lora(self, base_pipe_output, text_lora_config):
    if not self.supports_text_encoder_loras:
        pytest.skip("Text encoder LoRAs are not currently supported for this pipeline.")
    pipe = self.get_pipeline().to(torch_device)
    self.add_adapters_to_pipeline(pipe, text_lora_config)
    ...

# After
def test_simple_inference_with_text_lora(self, base_pipe_output):
    if "text_encoder" not in self.pipeline_class._lora_loadable_modules:
        pytest.skip("Text encoder LoRAs are not supported for this pipeline.")
    pipe = self.get_pipeline().to(torch_device)
    self.add_adapters_to_pipeline(pipe, components=["text_encoder"])
```

File: tests/pipelines/testing_utils/lora.py [hunk 9]
```diff
@@ -0,0 +1,1544 @@
...
+        if has_denoiser:
+            if "unet" in lora_loadable_modules and hasattr(pipe, "unet"):
+                modules_to_save["unet"] = pipe.unet
+
+            if "transformer" in lora_loadable_modules and hasattr(pipe, "transformer"):
+                modules_to_save["transformer"] = pipe.transformer
+
+        return modules_to_save
+
+    def _needs_text_encoder_lora_repair(self, pipe) -> bool:
```
Comment (DN6) [inline tests/pipelines/testing_utils/lora.py]: I'm a bit confused by this step. If there is an issue with `transformers>=5.6` we shouldn't work around and fix it inside the test?  Isn't this masking an issue? It means that currently loading  a LoRA for `CLIPTextModelWithProjection` will not work right?

File: tests/pipelines/testing_utils/lora.py [hunk 10]
```diff
@@ -0,0 +1,1544 @@
...
+            msg="Lora + 0 scale should lead to same result as no LoRA",
+        )
+
+        if "text_encoder" in self.pipeline_class._lora_loadable_modules:
+            text_encoder_root = getattr(pipe.text_encoder, "text_model", pipe.text_encoder)
+            assert text_encoder_root.encoder.layers[0].self_attn.q_proj.scaling["default"] == 1.0, (
+                "The scaling parameter has not been correctly restored!"
+            )
+
+    def test_simple_inference_with_text_lora_denoiser_fused(
```
Comment (DN6) [inline tests/pipelines/testing_utils/lora.py]: There are a few tests here that run the same fuse/unfuse check with minor variation. 
```
test_simple_inference_with_text_lora_fused
test_simple_inference_with_text_lora_denoiser_fused
test_simple_inference_with_text_denoiser_lora_unfused
```

If we use the the general add adapters logic suggested here:
https://github.com/huggingface/diffusers/pull/14268/changes#r3680162473

I think these tests can be collapsed into
```python
def test_lora_fuse_unfuse(self, base_pipe_output):
    pipe = self.get_pipeline().to(torch_device)
    adapted = self.add_adapters_to_pipeline(pipe)

    output_adapter = self.run_pipe(pipe)
    assert not torch.allclose(output_adapter, base_pipe_output, atol=1e-3, rtol=1e-3), (
        "LoRA should change the output"
    )

    pipe.fuse_lora(components=self.pipeline_class._lora_loadable_modules)
    assert pipe.num_fused_loras >= 1, f"{pipe.num_fused_loras=}, {pipe.fused_loras=}"
    for name, module in adapted.items():
        assert check_if_lora_correctly_set(module), f"Fusing should keep LoRA layers in {name}"

    output_fused = self.run_pipe(pipe)
    assert_tensors_close(
        output_fused,
        output_adapter,
        atol=1e-3,
        rtol=1e-3,
        msg="Fusing should not change the output",
    )

    pipe.unfuse_lora(components=self.pipeline_class._lora_loadable_modules)
    assert pipe.num_fused_loras == 0, f"{pipe.num_fused_loras=}, {pipe.fused_loras=}"
    for name, module in adapted.items():
        assert check_if_lora_correctly_set(module), f"Unfusing should keep LoRA layers in {name}"

    output_unfused = self.run_pipe(pipe)
    assert_tensors_close(
        output_unfused,
        output_adapter,
        atol=1e-3,
        rtol=1e-3,
        msg="Unfusing should restore the dynamic LoRA output",
    )
```

File: tests/models/testing_utils/lora.py (hunk 5 above)
Comment (sayakpaul) [reply tests/models/testing_utils/lora.py]: Yes but this is also interfaced in our model classes which is why the tests.

File: tests/models/transformers/test_models_transformer_sana.py (hunk 3 above)
Comment (sayakpaul) [reply tests/models/transformers/test_models_transformer_sana.py:122]: Because the conv kernel is missing for the matmuls needed on our CI runners.

**Comment by sayakpaul** [comment]
@DN6 very reasonable asks. I have addressed all of them along with some further improvements.

File: tests/pipelines/testing_utils/lora.py [hunk 11]
```diff
@@ -0,0 +1,1233 @@
...
+            out_no_lora = self.run_pipe(pipe)
+
+            assert not torch.allclose(out_lora, out_no_lora, atol=1e-3, rtol=1e-3)
+        finally:
+            # Clean up the hooks to prevent state leak
+            if hasattr(denoiser, "_diffusers_hook"):
+                denoiser._diffusers_hook.remove_hook(_GROUP_OFFLOADING, recurse=True)
+
+
+class UNetLoraTesterMixin(BaseLoraTesterMixin):
```
Comment (sayakpaul) [inline tests/pipelines/testing_utils/lora.py:1071]: Not used yet. But when we get to SD and SDXL, this will be leveraged.

File: tests/models/testing_utils/lora.py (hunk 5 above)
Comment (DN6) [reply tests/models/testing_utils/lora.py]: While that may be true, it would be better to invoke this functionality via diffusers API rather than calling `inject_adapter_in_model` / `set_peft_model_state_dict` directly.

Comment (sayakpaul) [reply tests/models/testing_utils/lora.py]: Agreed. Will push a fix.

Comment (sayakpaul) [reply tests/models/testing_utils/lora.py]: Fixed. Also pushed a similar change to `test_lora_B_bias`.

**Review by DN6 — APPROVED** [review]
